helix-term/src/handlers/completion.rs

```diff
@@ -291,6 +299,223 @@ fn request_completion(
+                        (Some(re_stem), Some(t)) => re_stem.starts_with(t),
+                        _ => true,
+                    };
+                if filename_starts_with_prefix {
+                    dir_entry.metadata().ok().map(|md| (path, md))
+                } else {
+                    None
+                }
+            })
+            .map(|(path, md)| {
+                let file_name = path
+                    .file_name()
+                    .unwrap_or_default()
+                    .to_string_lossy()
+                    .to_string();
+
+                let full_path = path.canonicalize().unwrap_or_default();
+                let full_path_name = full_path.to_string_lossy().into_owned();
+
+                let is_dir = full_path.is_dir();
+
+                let path_type = if md.is_symlink() {
+                    "link"
+                } else if is_dir {
+                    "folder"
+                } else {
+                    #[cfg(unix)]
+                    {
+                        use std::os::unix::fs::FileTypeExt;
+                        if md.file_type().is_block_device() {
+                            "block"
+                        } else if md.file_type().is_socket() {
+                            "socket"
+                        } else if md.file_type().is_char_device() {
+                            "character"
+                        } else if md.file_type().is_fifo() {
+                            "fifo"
+                        } else {
+                            "file"
+                        }
+                    }
+                    #[cfg(not(unix))]
+                    "file"
+                };
+
+                let resolved = if path_type == "link" { "resolved " } else { "" };
+
+                let documentation = Some(lsp::Documentation::MarkupContent(lsp::MarkupContent {
+                    kind: lsp::MarkupKind::Markdown,
+                    value: {
+                        #[cfg(unix)]
+                        {
+                            use std::os::unix::prelude::PermissionsExt;
+                            let mode = md.permissions().mode();
+
+                            let perms = [
+                                (libc::S_IRUSR, 'r'),
+                                (libc::S_IWUSR, 'w'),
+                                (libc::S_IXUSR, 'x'),
+                                (libc::S_IRGRP, 'r'),
+                                (libc::S_IWGRP, 'w'),
+                                (libc::S_IXGRP, 'x'),
+                                (libc::S_IROTH, 'r'),
+                                (libc::S_IWOTH, 'w'),
+                                (libc::S_IXOTH, 'x'),
+                            ]
+                            .iter()
+                            .fold(String::new(), |mut acc, (p, s)| {
+                                #[allow(clippy::unnecessary_cast)]
+                                acc.push(if mode & (*p as u32) > 0 { *s } else { '-' });
```

Comment:
```suggestion
                                acc.push(if mode & p > 0 { *s } else { '-' });
```

it looks like the compiler derefs this automatically now

Reply:
Actually I guess you could use `into_iter` instead of iter here and also drop the deref for `s`

Reply:
Yeah `into_iter` is possible, also added `String::with_capacity(9)` to avoid unnecessary allocs...

Btw. the cast is necessary for e.g. OSX (added a comment as well) as there is `p: u16`.